src/base/bittorrent/tracker.cpp

```diff
@@ -165,13 +178,13 @@ void Tracker::respondToAnnounceRequest()
       }*/
 
     // 3. Get port
-    if (!gets.contains("port")) {
+    if (!queryParams.contains("port")) {
         qDebug("Tracker: Missing port");
         status(103, "Missing port");
         return;
     }
     bool ok = false;
-    annonceReq.peer.port = gets.value("port").toInt(&ok);
+    annonceReq.peer.port = queryParams.value("port").toInt(&ok);
     if (!ok || annonceReq.peer.port < 1 || annonceReq.peer.port > 65535) {
```

Comment:
@arvidn, qBittorrent embedded Tracker implementation rejects Peers with `port=0` (legacy and very old behavior, so I don't know, why it was). But now I noticed that the libtorrent sends a couple of tracker requests, one with the correct port (which I set in the app settings), and the other, that is exactly the same, but with `port=0`. Please explain what this means and what the tracker should do.
@sledgehammer999, @Chocobo1, maybe you know something about it?

Reply:
libtorrent will announce with a 0-port if it does not have any listen sockets. For instance, if it's using a proxy and has set force-proxy (which will close the listen sockets, to close a back-door of directly connecting to the client). The issue is that if we connect to a tracker via a proxy, and could accept incoming uTP connections via a SOCKS5 proxy, that external port on the proxy is not known to us, so there is no way to post the correct port to the tracker. (in an ideal world, everyone would use UDP trackers and it would be extended to use the source port of the tracker connection itself to record the peer's listen port). So, posting our (local) listen port only leaks information about us, without providing any utility to the tracker or other peers.

Reply:
@arvidn, as I understand from your answer (and other discussions), if the tracker receives a request with "port=0", it should send a response with a port used in connection, right? I didn't find anything like that in the bittorrent protocol specs...

>libtorrent will announce with a 0-port if it does not have any listen sockets

But why it announce twice, one time with the given port (I set it listen to) and the second one with zero port?

Reply:
> as I understand from your answer (and other discussions), if the tracker receives a request with "port=0", it should send a response with a port used in connection, right?

I don't know what you mean. Are you're suggesting the tracker is supposed to respond with the client's source port for the tracker announce itself? That wouldn't be helpful since outgoing TCP connections are bound to an arbitrary source port.

> But why it announce twice, one time with the given port (I set it listen to) and the second one with zero port?

I don't know. I don't think it's supposed to. Can you catch the client doing that? with logs?

Reply:
>That wouldn't be helpful since outgoing TCP connections are bound to an arbitrary source port.

Oh, that's right... But what should the tracker do in this case? Sending the peer with zero port is meaningless too, isn't it? Just ignore such requests?..

Reply:
>Can you catch the client doing that? with logs?

This client is qBittorrent itself (i.e. linked libtorrent). I just added qBittorrent embedded tracker URL to one of my existing torrents and encountered error message. Then I debuged it and found that it announces twice.

Reply:
> Oh, that's right... But what should the tracker do in this case? Sending the peer with zero port is meaningless too, isn't it? Just ignore such requests?..

I think it's reasonable to respond to the request but not to add the peer to the peer list.

> This client is qBittorrent itself (i.e. linked libtorrent). I just added qBittorrent embedded tracker URL to one of my existing torrents and encountered error message. Then I debuged it and found that it announces twice.

Do you have logs or any other description of the scenario? Do you use a proxy for instance? which version of libtorrent are you using?
Multiple announces happens in libtorrent 1.1 when a tracker resolves to both an IPv4 and IPv6 address for instance. Is this over loopback by any chance? Then the question is why the port is inconsistent. Can you find any clues?